### PR TITLE
Ml/dialing and opening connections

### LINKED ISSUE #104: Dialing / opening connections
- Add transport/routing.go with [ShardInfo](https://github.com/scylladb/scylla-rust-driver/blob/ee16c710e6a7164550f43a13c7125dbe644326cd/scylla/src/routing.rs#L25)
- Add transport.ConnectionConfig for the 1st version we need
    * pub tcp_nodelay: bool,
    * pub connect_timeout: std::time::Duration,
    * pub default_consistency: Consistency,
- Add transport.[OpenConnection](https://github.com/scylladb/scylla-rust-driver/blob/ee16c710e6a7164550f43a13c7125dbe644326cd/scylla/src/transport/connection.rs#L1125) note that the OpenConnection function family should only do the dialing and Wrap the TCP connection. The protocol handshake shall be implemented as an init method called by `WrapConn` function.
- We should only support shard aware ports i.e. dialing with source port.

### PR DESCRIPTION
**Pull request by Michal-Leszczynski**
Transport: conn, created family of three OpenConn functions able to establish connection.
Transport: routing, added simple functions for randomly choosing specific kind of local port (Rust approach).

Fixes #104

### CODE REVIEW

**Comment by mmatczuk** [comment]
Cheers for detailed commit message, can you an empty line between header and body please.

File: transport/conn.go [hunk 1]
```diff
@@ -172,11 +174,61 @@ type Conn struct {
 	stream StreamIDAllocator
 }
 
+type ConnConfig struct {
+	tcpNoDelay bool
+	timeout    time.Duration
```
Comment (mmatczuk) [inline transport/conn.go:179]: This must be exported, start with T not t

File: transport/conn.go [hunk 2]
```diff
@@ -172,11 +174,61 @@ type Conn struct {
...
+	// defaultConsistency frame.Consistency
+}
+
 const (
 	requestChanSize = 1024
 	ioBufferSize    = 8192
 )
 
+// OpenShardConn opens connection mapped to a specific shard on scylla node.
+func OpenShardConn(rAddr string, si ShardInfo, cfg ConnConfig) (*Conn, error) { // nolint:unused // This will be used.
```
Comment (mmatczuk) [inline transport/conn.go:190]: I'd suggest and naming scheme change

- `rAddr` to `addr`
- `lAddr` to `localAddr`

File: transport/routing.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+const (
+	// Range of ports that can be used to establish connection.
+	maxPort = 65535
+	minPort = 49152
+)
+
+type ShardInfo struct {
+	shard    uint16
+	nrShards uint16
```
Comment (mmatczuk) [inline transport/routing.go:13]: Please export fields of exported structs.

File: transport/routing.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	return r/si.nrShards*si.nrShards + si.shard
+}
+
+// ShardPortIterator returns iterator for consecutive ports that are
+// mapped to a specific shard on scylla node.
+func ShardPortIterator(si ShardInfo) func() uint16 {
+	curr := RandomShardPort(si)
+
+	return func() uint16 {
+		curr += si.nrShards
```
Comment (mmatczuk) [inline transport/routing.go:33]: cur?

File: transport/conn.go [hunk 5]
```diff
@@ -172,11 +174,61 @@ type Conn struct {
...
+	conn, err := d.Dial("tcp", rAddr)
+	if err != nil {
+		return nil, fmt.Errorf("dialing TCP address %s: %w", rAddr, err)
+	}
+
+	tcpConn := conn.(*net.TCPConn)
+	if err = tcpConn.SetNoDelay(cfg.tcpNoDelay); err != nil {
+		return nil, fmt.Errorf("setting TCP no delay option: %w", err)
+	}
+	// TODO: add StreamIDAllocator.
```
Comment (mmatczuk) [inline transport/conn.go:228]: This will be removed.

File: transport/routing.go (hunk 4 above)
Comment (Michal-Leszczynski) [reply transport/routing.go:33]: It was supposed to mean current. Will currentPort be more descriptive?

File: docker-compose.yml [hunk 6]
```diff
@@ -6,3 +6,4 @@ services:
     command: ${SCYLLA_ARGS}
     ports:
       - "9042:9042"
+      - "19042:19042"
```
Comment (mmatczuk) [inline docker-compose.yml:9]: 👍

File: transport/routing.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+	maxRange := int(maxPort - si.NrShards + 1)
+	minRange := int(minPort + si.NrShards - 1)
+	r := uint16(rand.Intn(maxRange-minRange+1) + minRange)
+	return r/si.NrShards*si.NrShards + si.Shard
+}
+
+// ShardPortIterator returns iterator for consecutive ports that are
+// mapped to a specific shard on scylla node.
+func ShardPortIterator(si ShardInfo) func() uint16 {
+	currentPort := RandomShardPort(si)
```
Comment (mmatczuk) [inline transport/routing.go:30]: cur was fine port would be also fine.

Follow the rule of thumb that the length of var name represents its scope.
Here scope is small but not minuscule :semi

**Pull request merged**